### ISSUE TITLE
BlockStartParser not executed when markdown contains just one line

### ISSUE THREAD

**Issue by laacz** [thephpleague/commonmark#706]
**Version(s) affected**: 2.0.1

**Description**  
While trying to migrate over to 2.0, I've encountered an issue. My extension registers blockstart parser in order to replace youtube links on a single line with lite-youtube embeds. However, to my surprise block my custom parser is not called on that markdown.

**How to reproduce**  
Markdown (all on single line):

```markdown
https://youtube.com/?v=videoid
```

Barebones code example:

```php
<?php

use League\CommonMark\Node\Node;
use League\CommonMark\Renderer\ChildNodeRendererInterface;
use League\CommonMark\Environment\EnvironmentBuilderInterface;
use League\CommonMark\Extension\ExtensionInterface;
use League\CommonMark\Parser\Block\BlockStart;
use League\CommonMark\Parser\Block\BlockStartParserInterface;
use League\CommonMark\Parser\Cursor;
use League\CommonMark\Parser\MarkdownParserStateInterface;
use League\CommonMark\Renderer\NodeRendererInterface;

require('vendor/autoload.php');

class YoutubeLink {
    public function __construct($link) {
        die($link);
    }
}

class YoutubeLinksBlockStartParser implements BlockStartParserInterface
{
    public function tryStart(Cursor $cursor, MarkdownParserStateInterface $parserState): ?BlockStart
    {
        print_r(__CLASS__ . '::' . __METHOD__);
        die();
    }
}

class YoutubeLinksRenderer implements NodeRendererInterface {
    public function render(Node $node, ChildNodeRendererInterface $childRenderer) {
        print_r(__CLASS__ . '::' . __METHOD__);
        die();
    }
}


class MyMarkdownExtension implements ExtensionInterface
{
    public function register(EnvironmentBuilderInterface $environment): void
    {
        $environment
            ->addBlockStartParser(new YoutubeLinksBlockStartParser())
            ->addRenderer(YoutubeLink::class, new YoutubeLinksRenderer());
    }
}

use League\CommonMark\Environment\Environment;
use League\CommonMark\Extension\CommonMark\CommonMarkCoreExtension;
use League\CommonMark\MarkdownConverter;

$environment = new Environment();
$environment->addExtension(new CommonMarkCoreExtension());
$environment->addExtension(new MyMarkdownExtension());

$converter = new MarkdownConverter($environment);
echo $converter->convertToHtml("https://youtube.com/?v=videoid");
```

**Comment by colinodell** [comment]
This appears to be happening due to a new optimization added in 2.0.0:

https://github.com/thephpleague/commonmark/blob/1f3f61e9817ebeb55e15b3684a9b1ccdcb81e10c/src/Parser/MarkdownParser.php#L139-L141

Basically, we're assuming that all block indicators will be non-letter characters.  But in this case, you have a block starting with the letter `h`, and therefore this logic triggers the parser to avoid checking for block starts.

I'll have to do some performance profiling to determine what the potential impact would be over reverting this optimization.

**Comment by colinodell** [comment]
Undoing this particular optimization hurts performance by ~10% in my benchmarks, so I'm very hesitant to remove it, but I think I have a solution that should work for both of us... working on the implementation now!

**Comment by colinodell** [comment]
Fixed in version [2.0.2](https://github.com/thephpleague/commonmark/releases/tag/2.0.2).

Just make sure you call `->addBlockStartParser(new YoutubeLinksBlockStartParser())` with a priority of 250 or higher.

**Comment by laacz** [comment]
Thank you. Migration from 1.x to 2.x was no small feat. Took me two days to refactor on a small project. Hope 3.x won't break stuff this much :)
